Dccrg update

markusbattarbee opened the pull request:
Updates Vlasiator to work with DCCRG master (as of 26th May 2020)

Ran testpackage
Mostly zero diffs

Magnetosphere_polar_small
proton/vg_v_nonthermal_0                4.65661e-10                 5.82336e-16    
proton/vg_v_nonthermal_1                1.09139e-11                 1.86547e-16    
proton/vg_v_nonthermal_2                1.16415e-10                 2.44615e-16    
proton/vg_ptensor_nonthermal_diagonal_0                2.22143e-27                 8.72321e-16    
proton/vg_ptensor_nonthermal_diagonal_1                1.61559e-27                 3.21758e-16

ykempf commented:
Surprisingly little changes then. Any insight about how stable this is now in DCCRG? Will it change again any time soon?

ykempf commented:
Little birds confirm this will be stable, good.
But: should there not be some more elaborate AMR test, as this is used only in AMR code (except the () addition)?

ursg commented on vlasovsolver/cpu_trans_map_amr.cpp:
```diff
@@ -1364,8 +1364,11 @@ int get_sibling_index(dccrg::Dccrg<SpatialCell,dccrg::Cartesian_Geometry>& mpiGr
+
+   // get_all_children returns an array instead of a vector now, need to map it to a vector for find and distance
+   std::array<uint64_t, 8> siblingarr = mpiGrid.mapping.get_all_children(parent);
+   vector<CellID> siblings(&siblingarr[0], &siblingarr[0] + sizeof(siblingarr) / sizeof(siblingarr[0]));
```
That took me a moment to parse. Wouldn't ```siblangarr.begin(), siblingarr.end()``` fulfill the same function and be more understandable?

markusbattarbee replied:
Well, perhaps &siblingarr[0] can be replaced with siblingarr.begin()
but I tried
    vector<CellID> siblings(siblingarr.begin(),siblingarr.end());
and it gives major diffs. All the examples of generating a vector from an array used the method I used, apparently for a good reason.

markusbattarbee replied:
replacing &siblingarr[0] with siblingarr.begin() also gave diffs.

markusbattarbee replied:
> Little birds confirm this will be stable, good.
> But: should there not be some more elaborate AMR test, as this is used only in AMR code (except the () addition)?

Yeah, AMR testing underway and causing funny stuff. Don't merge yet.

ursg replied:
Bizarre. I guess my C++ kungfoo is weak today. Well, leave it as it is then.

markusbattarbee commented:
In AMR tests, there are some diffs. My theory:

I took regular Vlasiator dev and added a error check output here:
https://github.com/fmihpc/vlasiator/compare/dev...markusbattarbee:dev_translation_printouts

and got lots of warnings. I think this is triggered in this case:

---------------------------------------------------------------
|      L0        |      *L0*      |  *L1*  |   L1   |    L1   |
---------------------------------------------------------------

The starred L1 cell looks stencils cells from a 2-cell neighborhood, which includes the non-starred L1-cells and the starred L0 cell. However, if I understand correctly, it doesn't include the non-starred L0 cell. Thus, in the negative direction, it finds only one cell to put in the vlasov stencil source cells array, and replaces the NULL value (which should point to the non-starred L0) with a copy of the starred L0.

In my test to make Vlasiator work with DCCRG master, I switched to an iterative method for going through face neighbors, which should activate the non-starred source cell L0 correctly. In this one, I did not get warnings for a lack of neighbors for non-sysboundary cells. This would explain the diffs I was getting, and would classify as a bugfix.

Thoughts?

Still working on redoing some testpackage runs to be absolutely sure of what I'm seeing, though.

ursg commented:
Interesting. Do we still have the mini-app that was doing a 1-d propagation? I remember dimly from back then, that is tended to cause some artefacts lagging behind the gauss peak that it transported through, and now I'm wondering if this would fix those.

ykempf commented:
I believe the miniapps are stored in that directory in the repo.

markusbattarbee commented:
Yep vlasiator/mini-apps/translation_pencils/trans_test.cpp  could be used to check this.
